[hf inference] ASR remote inference model parser impl

Ankush-lastmile opened the pull request:
[hf inference] ASR remote inference model parser impl

Implementation of the HuggingFaceAutomaticSpeechRecognition Model parser using the inference endpoint to run inference. Python API takes in bytes as well as path, skip binary for now.

Very similar to #1018

## Testplan
<img width="1000" alt="Screenshot 2024-01-24 at 10 37 05 PM" src="https://github.com/lastmile-ai/aiconfig/assets/141073967/808956ce-e3be-4528-9f34-c8d31d704ddb">

1. Temporarily add model parser to Gradio Cookbook model parser registry.
```
    asr = HuggingFaceAutomaticSpeechRecognitionRemoteInference()
    AIConfigRuntime.register_model_parser(
        asr, asr.id()
    )
```

2. run AIConfig Edit on Gradio example

`python3 -m 'aiconfig.scripts.aiconfig_cli' edit --aiconfig-path=cookbooks/Gradio/huggingface.aiconfig.json --parsers-module-path=cookbooks/Gradio/hf_model_parsers.py --server-mode=debug_servers`

rholinshead commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/remote_inference_client/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,359 @@
+
+        Usage:
+        1. Create a new model parser object with the model ID of the model to use.
+                parser = HuggingFaceAutomaticSpeechRecognitionTransformer()
```
```suggestion
                parser = HuggingFaceAutomaticSpeechRecognitionRemoteInference()
```

Ankush-lastmile replied:
updated

rholinshead commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/remote_inference_client/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,359 @@
+
+        # For now, support Path as path to local image file or str as uri only
+        # TODO: Support bytes and BinaryIO
+        image: str = data["audio"]
```
```suggestion
        audio: Union[Path, str, bytes, BinaryIO] = data["audio"]
```

Should fix the name and typing here

rholinshead commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/remote_inference_client/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,359 @@
+        # without loading it, so just use the discrete type by default
+        mime_type = "audio"
+
+        if isinstance(image, Path):
```
here and throughout
```suggestion
        if isinstance(audio, Path):
```

Ankush-lastmile replied:
updated

rholinshead commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/remote_inference_client/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,359 @@
+
+        completion_data = refine_completion_params(model_settings)
+
+        # HF Python API supports input types of bytes, file path, and a dict containing raw sampled audio. Also supports multiple input
```
It only supports single audio input, based on https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/inference/_client.py#L304, right?

If so, please update the comment and variable handling here to not be an array

Ankush-lastmile replied:
updated comment

rholinshead commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/remote_inference_client/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,359 @@
+        # For now, support multiple or single uri's as input
+        inputs = validate_and_retrieve_audio_from_attachments(prompt)
+
+        if len(inputs) > 1:
+            raise ValueError(
+                f"Multiple audio inputs are not supported for the HF Automatic Speech Recognition Inference api. Please specify a single audio input attachment."
+            )
```
Ah, ok, that's what I was thinking. Instead of doing this, we should just make the validate_and_retrieve function return a single value, not array

Ankush-lastmile replied:
updated. refactpred validate_and_retrieve  returns a single value, not an array

rholinshead commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/remote_inference_client/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,359 @@
+            
+            # HuggingFace Automatic Speech Recognition outputs should only ever be string
+            # format so shouldn't get here, but just being safe
+            return json.dumps(output_data, indent=2)
```
Based on your comment on the image_2_text one, maybe this should raise a ValueError instead?

Ankush-lastmile replied:
Yep, updated. Had previously copy pasted what you had for consistency.

rholinshead reviewed (changes requested):
Some minor changes, mostly from copy/paste

Ankush-lastmile commented:
- Fixed a couple of nits on comments
- Renamed any references of `image` to `audio`
- Refactored `validate_and_retrieve_audio_from_attachments` to also validate only one audio attachment

## Testplan,

Same as in original pr description, output had the same output so omitting the screenshot

rholinshead reviewed (approved):
Accepting to unblock, but please update the mimetype validation before landing

Ankush-lastmile commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/remote_inference_client/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,362 @@
+            f"Attachment has no mime type. Specify the audio mimetype in the aiconfig"
+        )
+
+    if not attachment.mime_type.startswith("audio/"):
```
ah I see what you mean. Technically this doesn't break anything yet but will be needed.

Updated, thanks for catching

Ankush-lastmile commented:
update the mimetype validation to check `audio`, not `audio/`
